finat/restricted.py

```diff
@@ -0,0 +1,219 @@
+
+
+@restrict.register(finat.DiscontinuousElement)
+def restrict_discontinuous(element, domain, take_closure):
+    if domain == "interior":
+        return element
+    else:
+        return null_element
+
+
+@restrict.register(finat.EnrichedElement)
+def restrict_enriched(element, domain, take_closure):
+    elements = tuple(restrict(e, domain, take_closure)
+                     for e in element.elements)
+    elements = tuple(e for e in elements if e is not null_element)
+    if elements:
+        return finat.EnrichedElement(elements)
+    else:
+        return null_element
+
+
+@restrict.register(finat.HCurlElement)
+def restrict_hcurl(element, domain, take_closure):
+    restricted = restrict(element.wrappee, domain, take_closure)
+    if restricted is null_element:
+        return null_element
+    else:
+        if isinstance(restricted, finat.EnrichedElement):
+            return finat.EnrichedElement(finat.HCurlElement(e)
+                                         for e in restricted.elements)
+        else:
+            return finat.HCurlElement(restricted)
+
+
+@restrict.register(finat.HDivElement)
+def restrict_hdiv(element, domain, take_closure):
+    restricted = restrict(element.wrappee, domain, take_closure)
+    if restricted is null_element:
+        return null_element
+    else:
+        if isinstance(restricted, finat.EnrichedElement):
+            return finat.EnrichedElement(finat.HDivElement(e)
+                                         for e in restricted.elements)
+        else:
+            return finat.HDivElement(restricted)
+
+
+@restrict.register(finat.mixed.MixedSubElement)
+def restrict_mixed(element, domain, take_closure):
+    raise NotImplementedError("Don't know how to handle mixed + restricted")
+
+
+@restrict.register(finat.GaussLobattoLegendre)
+def restrict_gll(element, domain, take_closure):
+    try:
+        return FiatElement(FIAT.RestrictedElement(element._element, restriction_domain=domain))
+    except ValueError:
+        return null_element
+
+
+@restrict.register(finat.GaussLegendre)
+def restrict_gl(element, domain, take_closure):
+    if domain == "interior":
+        return element
+    else:
+        return null_element
+
+
+def r_to_codim(restriction, dim):
+    if restriction == "interior":
+        return 0
+    elif restriction == "facet":
+        return 1
+    elif restriction == "face":
+        return dim - 2
+    elif restriction == "edge":
+        return dim - 1
+    elif restriction == "vertex":
+        return dim
+    else:
+        raise ValueError
+
+
+def codim_to_r(codim, dim):
+    d = dim - codim
+    if codim == 0:
+        return "interior"
+    elif codim == 1:
+        return "facet"
+    elif d == 0:
+        return "vertex"
+    elif d == 1:
+        return "edge"
+    elif d == 2:
+        return "face"
+    else:
+        raise ValueError
+
+
+@restrict.register(finat.TensorProductElement)
+def restrict_tpe(element, domain, take_closure):
```

Comment:
It would be nice to add a comment here explaining the algorithm.

Reply:
Done, hopefully makes sense.

Reply:
Yep, looks great.